synapse_antispam/mjolnir/antispam.py

```diff
@@ -14,115 +14,145 @@
-        self.block_usernames = config.get("block_usernames", False)
-        self.list_room_ids = config.get("ban_lists", [])
-        self.rooms_to_lists = {}  # type: Dict[str, BanList]
-        self.api = api
-
-        # Now we build the ban lists so we can match them
-        self.build_lists()
-
-    def build_lists(self):
-        for room_id in self.list_room_ids:
-            self.build_list(room_id)
-
-    def build_list(self, room_id):
-        logger.info("Rebuilding ban list for %s" % (room_id))
-        self.get_list_for_room(room_id).build()
-
-    def get_list_for_room(self, room_id):
-        if room_id not in self.rooms_to_lists:
-            self.rooms_to_lists[room_id] = BanList(api=self.api, room_id=room_id)
-        return self.rooms_to_lists[room_id]
-
-    def is_user_banned(self, user_id):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.user_rules:
-                if rule.matches(user_id):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
-
-    def is_room_banned(self, invite_room_id):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.room_rules:
-                if rule.matches(invite_room_id):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
-
-    def is_server_banned(self, server_name):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.server_rules:
-                if rule.matches(server_name):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
+        self._api = api
+        self._config = config
+
+        # Start a timer to rip the new rules off the server. We hit the rule server
+        # often to avoid missing rules - the response should be tens of milliseconds
+        # in duration, though network issues are possible - Twisted should automatically
+        # stack the calls for us.
+        # HACK: Private member access (_hs)
+        api._hs.get_clock().looping_call(self._update_rules, 5 * 1000)
+
+        # These are all arrays of compile()'d code from the ban list server.
+        # First match wins. See _update_rules() for more info.
+        self._code_spam_checks = []
+        self._code_invite_checks = []
+        self._code_profile_checks = []
+        self._code_create_room_checks = []
+        self._code_create_alias_checks = []
+        self._code_publish_room_checks = []
+
+    def _update_rules(self):
+        async def run():
+            try:
+                logger.info("Updating mjolnir rules...")
+
+                # HACK: Private member access (_hs)
+                resp = await self._api._hs.get_proxied_http_client().get_json(self._config['rules_url'])
+
+                # *** !! DANGER !! ***
```

Comment:
The Python module looks generally sane and I don't see anything obviously wrong.

From a defence in depth perspective, I do have a general unease around using `eval` on code received from a server. I understand this is coming from a trusted server, but it does make it easier for an attacker to escalate and move laterally in case he somehow manages to take control of the rules server.

Is there a particular reason why the rules need to be Python code instead of a small (e.g. JSON-based) DSL that is only able to encode the primitives that will reasonably be used for constructing ban expressions?

Reply:
I agree with @dkasak that this knowingly leaves open a potential insecure vector. Note that the potentially generated rules (e.g. `resp["checks"][...]["search"]`) seems to be one of:

* `event.get(..., '')`
* `user_profile[...]`
* A direct variable (e.g. `room_id` or `user_id`).
* `UserID.from_string(user_id).domain`

Some potential alternatives:

* Have the spam checker call the server for each query and handle a simple yes/no response. 🤷 This would only work for Synapse v1.25.0 where spam checker methods can be async and might have serious performance implications.
* Since the generated expressions are generally accessing variables or dictionary keys, have the server return a dotted dictionary accessor (e.g. `user_id` or `user_profile.user_id` and do some magic with splitting on `.`, something like:

```python
accessor = "user_profile.user_id"  # this would be returned by the server
value = {"user_profile": user_profile}
for key in accessor.split("."):
    value = getattr(value, key, "")
if re.search(check["pattern"], value):
    return True
```

Note that something special would need to be done to get the user's domain, but that could be handled fairly easily by doing something like the following for the initial value:

```python
value = {
    "user_id": user_id,
    "user_domain": UserID.from_string(user_id).domain,
}
```

The accessor would then be `user_id` or `user_domain`.

In general this approach seems overly flexible for the rules that get generated in `RuleServer.ts`, but maybe there's a future plan I don't know about. Even my example above seems like it could be simplified quite a bit by knowing the data returned for each rule and what it gets applied against, which would also make the code a bit less abstract.

Reply:
Note that if the dotted accessor approach is taken, the accessor keys should still be checked against a whitelist instead of allowing everything (which brings us back to a DSL). If a fully general accessor is allowed, it might be possible for an attacker to navigate the object to a secret string and then leak it by successively adapting a rule and checking whether it triggers.